### PR TITLE
Generate: SinkCache can handle iterative prompts

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes the case where `SinkCache` is used in a chat bot, receiving new prompts after giving an answer. Fix developed with @tomaarsen 

Here's an example of a script that works after this PR:
```py
from transformers import AutoTokenizer, SinkCache, AutoModelForCausalLM, TextStreamer
import torch
from datasets import load_dataset

# Loading the model & tokenizer
model_id = "HuggingFaceH4/zephyr-7b-beta"
model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", torch_dtype=torch.float16)
tokenizer = AutoTokenizer.from_pretrained(model_id)

# Loading the prompts to simulate user interactions
prompt_dataset = load_dataset("HuggingFaceH4/mt_bench_prompts", split="train")
prompts = [prompt for prompts in prompt_dataset["prompt"] for prompt in prompts]

# Prepare generation settings
cache = SinkCache(window_length=1024, num_sink_tokens=4)
streamer = TextStreamer(tokenizer)

input_ids = torch.tensor([], device=model.device, dtype=torch.int)
for prompt in prompts:
    # Tokenize the prompt with the correct chat template
    chat = [{"role": "user", "content": prompt}]
    input_ids = torch.cat((input_ids, tokenizer.apply_chat_template(chat, return_tensors="pt", add_generation_prompt=True).to(model.device)), dim=1)
    # input_ids = tokenizer.apply_chat_template(chat, return_tensors="pt").to(model.device)

    # Perform the generation
    gen_out = model.generate(input_ids, do_sample=False, max_new_tokens=100, past_key_values=cache, use_cache=True, streamer=streamer)

    # input_ids = torch.cat((input_ids, gen_out), dim=1)
    input_ids = gen_out

    # If desired, decode the output from this prompt
    decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
```

### CODE REVIEW

File: src/transformers/cache_utils.py [hunk 1]
```diff
@@ -239,8 +250,8 @@ def update(
         """
         # Optional kwargs for `SinkCache` -- needed on models using RoPE. `partial_rotation_size` is used on models
         # with partially rotated position embeddings, like Phi or Persimmon.
-        sin = cache_kwargs.get("sin")
-        cos = cache_kwargs.get("cos")
+        sin = cache_kwargs.get("sin")[: self.window_length]
```
Comment (gante) [inline src/transformers/cache_utils.py:253]: Slicing here is needed if more than one token is fed at once, after the cache is full.

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -405,6 +405,12 @@ def forward(
         if past_key_value is not None:
             cache_kwargs = {"sin": sin, "cos": cos}  # Specific to RoPE models
             key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
+            # If the cache has a fixed length and we were about to go beyond it, update the key/value length and the
+            # attention mask accordingly. `.update()` handles the cache cropping internally if needed.
+            kv_max_length = past_key_value.get_max_length()
+            if kv_max_length is not None and kv_seq_len > kv_max_length:
+                kv_seq_len = kv_max_length
+                attention_mask = attention_mask[:, :, :, -kv_seq_len:]
```
Comment (gante) [inline src/transformers/models/llama/modeling_llama.py:413]: The attention mask must be sliced to match the length of `key_states`, which might have been sliced in `.update()` in fixed-length caches

**Comment by tomaarsen** [comment]
Some additional results @gante 

image

* `transformers_old` refers to before #27700 (thanks @kevinhu for the change!)
* `transformers_56be5e80` refers to commit 56be5e80, i.e. near main
* `transformers_attn_sink_1024_4_pr-27907` refers to this PR, with `SinkCache(1024, 4)`

This experiment is with calling the models with individual tokens exclusively. Using `SinkCache` makes the memory usage linear at a very low cost in perplexity.

---

Additionally, I'm doing a test with calling Mistral-7B in this case with this PR using multiple tokens at once. I took indices 0 to 63k of a book from pg19, only kept 20% of all indices, and then fed the model with the tokens between subsequent indices. The running cache is also included. The NLL loss is then converted to perplexity.
**Note:** We can't compare this with the perplexities from the previous graph: we should only try and observe whether the model eventually increases in perplexity.

image

The same script crashes on `main`.

I think this works & it works well. 

- Tom Aarsen

**Review by tomaarsen — APPROVED** [review]

File: src/transformers/cache_utils.py (hunk 1 above)
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:253]:
```suggestion
        sin = cache_kwargs.get("sin")[-self.window_length:]
```
would that not make more sense? since that's the side we split the cache from no?

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:413]: Ouch not a fan of that. 
We have something similar for the mistral sliced window but not in favor of keeping this. That should either go in the attention convert, which should slice it, or in the cache_kwargs as it's clearly sink cache and window cache specific 😉

**Review by ArthurZucker — COMMENTED** [review]
Thanks

File: src/transformers/cache_utils.py (hunk 1 above)
Comment (tomaarsen) [reply src/transformers/cache_utils.py:253]: I'll run some tests for this

Comment (tomaarsen) [reply src/transformers/cache_utils.py:253]: Updated my previous message with my findings

Comment (gante) [reply src/transformers/cache_utils.py:253]: @ArthurZucker @tomaarsen that's the beauty of sin and cos: the rerotation is based on the relative angles, and the relative angles are the same regardless of the side we slice 🙌 

If you place a debugger, you can see that the sliced tensors are different, but the rerotation coefficients are exactly the same!

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/llama/modeling_llama.py:413]: The alternative would be to make `prepare_inputs_for_generation` to prepare the sliced mask in advance (which is how TF/JAX do it). Going to check if it is feasible

**Comment by gante** [comment]
@tomaarsen no, the script in the PR header runs endlessly without issues 🤔 LMK if you can find a reproducer

**Comment by gante** [comment]
@tomaarsen should be fixed

**Comment by tomaarsen** [comment]
@gante Works great now!
image

Red is the baseline, I can only run it to about ~15k seq length until my PC completely freezes.

File: src/transformers/models/mistral/modeling_mistral.py [hunk 3]
```diff
@@ -268,7 +268,7 @@ def forward(
                     "for auto-regressive decoding with k/v caching, please make sure to initialize the attention class "
                     "with a layer index."
                 )
-            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
+            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:271]: slice_length or window_length might be better? but a nit feel free to ignore

**Review by ArthurZucker — APPROVED** [review]
Thanks for the prompt update

**Comment by YJHMITWEB** [comment]
@gante @tomaarsen Thanks!
